source/common/http/headers.h

```diff
@@ -5,15 +5,40 @@
 #include "envoy/http/header_map.h"
 
 #include "common/singleton/const_singleton.h"
+#include "common/singleton/threadsafe_singleton.h"
 
 namespace Envoy {
 namespace Http {
 
+// This class allows early override of the x-envoy prefix from bootstrap config,
+// so that servers can configure their own x-custom-string prefix.
+//
+// Once the HeaderValues const singleton has been created, changing the prefix
+// will have no further effect.
+class PrefixValue {
+public:
+  const char* prefix() {
+    absl::ReaderMutexLock lock(&m_);
+    return prefix_;
+  }
+  void setPrefix(const char* prefix) {
+    absl::WriterMutexLock lock(&m_);
+    prefix_ = prefix;
+  }
+
+private:
+  const char* default_prefix_ = "x-envoy";
+  const char* prefix_ = default_prefix_;
```

Comment:
Is it possible to make this an absl::optional such that trying to read the prefix before it is set will crash? Essentially I'm wondering if we can set in all cases and then crash if we read before set. This way we are sure we don't have inconsistent values somewhere.

Another way to do this is to independently track when it is read, and then crash if the prefix is set after it has ever been read?

Reply:
We can, but I don't think it gets us anything other than a billion test changes.  Right now we verify that in the one place we set it, that it gets set correctly, so if anyone used the header map beforehand and the values don't match we will crash at server start up.

If we go the route you suggest, we'll have to set the singleton in all the unit tests and integration tests which use headers and not server.cc which I think is more trouble than it's worth given we have crash-based validation either way.  WDYT?

Reply:
I'm just worried about this leaking outside of headers somehow and we don't notice an inconsistency which seems pretty bad. 

Would it require test changes if we did my second suggestion which is to see if the value has ever been read before it is set? I think that would require no changes to any tests, since it would only fail in the case where we read before set.

Reply:
Oh sorry, I think that'll break less, possibly only the new integration test and if so I can work around.
looking